src/build_system.ml

```diff
@@ -1528,29 +1552,34 @@ let package_deps t pkg files =
         acc
       else begin
         rules_seen := Rule.Id.Set.add !rules_seen ir.id;
-        let _, dyn_deps =
-          match ir.exec with
-          | Running { rule_evaluation; _ }
-          | Evaluating_rule { rule_evaluation; _ } ->
-            Option.value_exn (Fiber.Future.peek rule_evaluation)
-          | Not_started _ -> assert false
-        in
-        let action_deps =
-          Static_deps.action_deps
-            (Lazy.force ir.static_deps)
-        in
-        Path.Set.fold
-          (Deps.path_union action_deps dyn_deps)
-          ~init:acc ~f:loop
+        match Fiber.Once.peek ir.static_deps with
+        | None ->
+          (* We know that at this point of execution, all the relevant
+             ivars have been filled *)
+          assert false
+        | Some static_deps ->
+          let rule_deps = Static_deps.rule_deps static_deps in
+          match Rule_fn.peek (Fdecl.get t.build_rule_def) ir with
+          | None -> assert false
+          | Some (_, deps) ->
+            let dyn_deps = Deps.path_diff deps rule_deps in
+            let action_deps =
+              Static_deps.action_deps static_deps |> Deps.paths
+            in
+            Path.Set.fold (Path.Set.union action_deps dyn_deps)
+              ~init:acc ~f:loop
       end
   in
   let open Build.O in
   Build.paths_for_rule files >>^ fun () ->
   (* We know that at this point of execution, all the relevant ivars
      have been filled *)
-  Path.Set.fold files ~init:Package.Name.Set.empty ~f:loop_deps
+  Path.Set.fold files ~init:Package.Name.Set.empty
+    ~f:(fun f acc -> loop_deps f acc)
```

Comment:
Curious why this change was made. The eta reduced version seemed preferable to me

Reply:
Oops, it's just a left-over. Some code was added and then removed, but it wasn't simplified afterwards.